SignalMessaging/attachments/ApproveContactShareViewController.swift

```diff
@@ -0,0 +1,556 @@
+
+    var fieldViews = [ContactShareFieldView]()
+
+    var nameLabel: UILabel!
+
+    // MARK: Initializers
+
+    @available(*, unavailable, message:"use other constructor instead.")
+    required public init?(coder aDecoder: NSCoder) {
+        fatalError("unimplemented")
+    }
+
+    @objc
+    required public init(contactShare: OWSContact, contactsManager: OWSContactsManager, delegate: ApproveContactShareViewControllerDelegate) {
+        self.contactsManager = contactsManager
+        self.contactShare = contactShare
+        self.delegate = delegate
+
+        super.init(nibName: nil, bundle: nil)
+
+        buildFields()
+    }
+
+    func buildFields() {
+        var fieldViews = [ContactShareFieldView]()
+
+        // TODO: Avatar
+
+        for phoneNumber in contactShare.phoneNumbers {
+            let field = ContactSharePhoneNumber(phoneNumber)
+            let fieldView = ContactShareFieldView(field: field, previewViewBlock: { [weak self] _ in
+                guard let strongSelf = self else { return UIView() }
+                return strongSelf.previewView(forPhoneNumber: phoneNumber)
+            })
+            fieldViews.append(fieldView)
+        }
+        for email in contactShare.emails {
+            let field = ContactShareEmail(email)
+            let fieldView = ContactShareFieldView(field: field, previewViewBlock: { [weak self] _ in
+                guard let strongSelf = self else { return UIView() }
+                return strongSelf.previewView(forEmail: email)
+            })
+            fieldViews.append(fieldView)
+        }
+        for address in contactShare.addresses {
+            let field = ContactShareAddress(address)
+            let fieldView = ContactShareFieldView(field: field, previewViewBlock: { [weak self] _ in
+                guard let strongSelf = self else { return UIView() }
+                return strongSelf.previewView(forAddress: address)
+            })
+            fieldViews.append(fieldView)
+        }
+
+        self.fieldViews = fieldViews
+    }
+
+    override public var canBecomeFirstResponder: Bool {
+        return true
+    }
+
+    // MARK: - View Lifecycle
+
+    override public func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        updateNavigationBar()
+    }
+
+    override public func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+    }
+
+    override public func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+    }
+
+    override public func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+    }
+
+    override public func loadView() {
+        super.loadView()
+
+        self.navigationItem.title = NSLocalizedString("CONTACT_SHARE_APPROVAL_VIEW_TITLE",
+                                                      comment: "Title for the 'Approve contact share' view.")
+
+        self.view.preservesSuperviewLayoutMargins = false
```

Comment:
I often see you ignoring the superview layout margins, but I'm not sure why. Don't we generally want to respect the system layout margins? 

i.e. when do we want to pin to the margin, it seems like we'd almost never want to be closer than 16px from the screen edge.

If what you want instead is to pin a view to the edge, and not the margin, then pin to the edge, rather than setting the margin to zero and pinning to the margin.

Reply:
Sorry, based on previous PR feedback I thought you preferred always pinning to margins and not edges.  I must have misunderstood.  

Note that AFAIK the correct way to pin subviews of a `UIScrollView` to the "full width" of the scroll view is to use the `layoutMarginsGuide`, via `autoPinLeadingToSuperviewMargin` for example.